### PR TITLE
[Frontend] Split jax_extras into several modules

### PR DESCRIPTION
**Pull request by grwlf**
This is an utility PR split `./utils/jax_extras.py` into three sub-modules. The goal is to make a complex https://github.com/PennyLaneAI/catalyst/pull/370 PR simpler.

### CODE REVIEW

File: frontend/catalyst/jax_extras/patches.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+        return x
+    elif isinstance(x, Tracer):
+        return x.aval
+    else:
+        return concrete_aval(x)
+
+
+def _no_clean_up_dead_vars(_eqn, _env, _last_used):
+    """A stub to workaround the Jax ``KeyError 'a'`` bug during the lowering of Jaxpr programs to
+    MLIR with the dynamic API enabled."""
```
Comment (erick-xanadu) [inline frontend/catalyst/jax_extras/patches.py:50]:
```suggestion
    MLIR with the dynamic API enabled.
    Relevant upstream ticket: https://github.com/google/jax/issues/18936"""
```

File: frontend/catalyst/jax_extras/patches.py [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+    return None
+
+
+def _gather_shape_rule_dynamic(
```
Comment (erick-xanadu) [inline frontend/catalyst/jax_extras/patches.py:54]: @grwlf did you open an issue for this upstream?

Comment (grwlf) [reply frontend/catalyst/jax_extras/patches.py:55]: Yes, this one https://github.com/google/jax/pull/19083

File: frontend/catalyst/jax_extras/patches.py [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+):  # pragma: no cover
+    """Validates the well-formedness of the arguments to Gather. Compared to the original version,
+    this implementation skips static shape checks if variable dimensions are used.
+
+    This function has been modified from its original form in the JAX project at
+    https://github.com/google/jax/blob/88a60b808c1f91260cc9e75b9aa2508aae5bc9f9/jax/_src/lax/slicing.py#L1438
+    version released under the Apache License, Version 2.0, with the following copyright notice:
+
+    Copyright 2021 The JAX Authors.
+    """
```
Comment (erick-xanadu) [inline frontend/catalyst/jax_extras/patches.py:74]:
```suggestion
TODO(@grwlf): delete once PR https://github.com/google/jax/pull/19083 has been merged
```

**Review by erick-xanadu — APPROVED** [review]
Just add a reference to the PR so that it can be visible in the code.

Thanks @grwlf!

File: frontend/catalyst/jax_extras/patches.py (hunk 2 above)
Comment (erick-xanadu) [reply frontend/catalyst/jax_extras/patches.py:55]: Just add a reference to this PR in the code please :)

**Pull request merged**